### ISSUE TITLE
scaled_dot_product_attention produces incorrect results with attn_mask on large sequence lengths

### ISSUE THREAD

**Issue by Luo-Yihang** [pytorch/pytorch#162588]
### 🐛 Describe the bug

`torch.nn.functional.scaled_dot_product_attention` produces incorrect results when the input is with **large sequence lengths** and using **attn_mask**.

As shown in the example, when the token length exceeds `100,000` and an `attn_mask` is applied, `scaled_dot_product_attention` produces results that are misaligned compared to a naive implementation (without using the mask).

This seems to be caused by numerical inaccuracies accumulating inside the `scaled_dot_product_attention` implementation. This should be fixed.

Example code _(This run may use around 70GB of memory. To reduce usage when reproducing, please increase the number of groups or use bfloat16)_:

```
import argparse
from typing import Tuple

import torch
import torch.nn.functional as F


@torch.no_grad()
def run_once(
    L: int,
    S: int,
    d: int,
    dtype: torch.dtype,
    device: torch.device,
    groups: int,
) -> Tuple[float, float]:
    # torch.manual_seed(0)

    B, H = 1, 16
    x = torch.randn(B, H, L, d, device=device, dtype=dtype) * 10

    attn_mask = torch.full((L, S), float("-inf"), device=device, dtype=dtype)
    q_bounds = [int(x) for x in torch.linspace(0, L, groups + 1, device=device).round().tolist()]
    for i in range(groups):
        q0, q1 = q_bounds[i], q_bounds[i + 1]
        attn_mask[q0:q1, :q1] = 0

    out_fused = x

    out_fused = F.scaled_dot_product_attention(x, x, x, attn_mask=attn_mask)

    outs = []
    for i in range(groups):
        q0, q1 = q_bounds[i], q_bounds[i + 1]
        xi_q = x[..., q0:q1, :]
        xi_k = x[..., :q1, :]
        xi_v = x[..., :q1, :]
        oi = F.scaled_dot_product_attention(xi_q, xi_k, xi_v)
        outs.append(oi)
    out_naive = torch.cat(outs, dim=-2)

    diff = (out_fused - out_naive).abs()
    mean_abs = diff.mean().item()
    denom = out_naive.abs().clamp_min(1e-12)
    mean_rel = (diff / denom).mean().item()

    # Print last sample_n samples for inspection
    sample_n = 4
    fused_sample = out_fused.reshape(-1)[-sample_n:].cpu().tolist()
    naive_sample = out_naive.reshape(-1)[-sample_n:].cpu().tolist()
    diff_sample = diff.reshape(-1)[-sample_n:].cpu().tolist()
    print(f"fused_sample={fused_sample}")
    print(f"naive_sample={naive_sample}")
    print(f"diff_sample={diff_sample}")

    return mean_abs, mean_rel


def main():
    parser = argparse.ArgumentParser(description="Compare SDPA vs naive on CUDA with additive mask.")
    parser.add_argument("--dtype", choices=["float32", "float16", "bfloat16"], default="float32")
    parser.add_argument("--groups", type=int, default=10)
    args = parser.parse_args()

    device = torch.device("cuda")

    dtype_map = {
        "float32": torch.float32,
        "float16": torch.float16,
        "bfloat16": torch.bfloat16,
    }
    dtype = dtype_map[args.dtype]

    print(f"dtype={args.dtype} groups={args.groups}")
    print("L,S,mean_abs_err,mean_rel_err")

    lengths = [1000, 10000, 50000, 100000]
    for L in lengths:
        S = L
        try:
            mean_abs, mean_rel = run_once(L=L, S=S, d=64, dtype=dtype, device=device, groups=args.groups)
            print(f"{L},{S},{mean_abs:.3e},{mean_rel:.3e}")
        except RuntimeError as e:
            print(f"{L},{S},OOM/ERR,{str(e).splitlines()[0]}")


if __name__ == "__main__":
    main()
```

Output:

```
dtype=float32 groups=10
L,S,mean_abs_err,mean_rel_err
fused_sample=[13.806726455688477, -2.0214109420776367, -1.819523572921753, 6.879058361053467]
naive_sample=[13.806726455688477, -2.0214109420776367, -1.819523572921753, 6.879058361053467]
diff_sample=[0.0, 0.0, 0.0, 0.0]
1000,1000,0.000e+00,0.000e+00
fused_sample=[17.007545471191406, -10.924842834472656, 11.69533634185791, 18.47803497314453]
naive_sample=[17.007545471191406, -10.924842834472656, 11.69533634185791, 18.47803497314453]
diff_sample=[0.0, 0.0, 0.0, 0.0]
10000,10000,0.000e+00,0.000e+00
fused_sample=[-12.53640365600586, -11.38840103149414, 2.667792320251465, -14.007373809814453]
naive_sample=[-12.53640365600586, -11.38840103149414, 2.667792320251465, -14.007373809814453]
diff_sample=[0.0, 0.0, 0.0, 0.0]
50000,50000,0.000e+00,0.000e+00
fused_sample=[22.241771697998047, 15.727160453796387, 16.34319496154785, -0.41784247756004333]
naive_sample=[12.36245346069336, 5.561803817749023, -5.277179718017578, 4.265460968017578]
diff_sample=[9.879318237304688, 10.165356636047363, 21.62037467956543, 4.683303356170654]
100000,100000,3.361e+00,6.941e+05
```

### Versions

Tested on Torch 2.7.1, though other versions may also be affected.

**Comment by yi-kai-wang** [comment]
# Minimal repro: SDPA (efficient backend) silently wrong for seq length > 2^16, with a discrete step exactly at 65536, independent of head_dim

This is a tighter reproduction of the "incorrect results with `attn_mask` on large
sequence lengths" behavior. It pins the failure to a **discrete step at exactly
`2**16 = 65536`** and shows the boundary is **independent of `head_dim`**, which
argues the cause is a **seq-dimension 16-bit index/stride overflow inside the
efficient-attention kernel**, not gradual numerical accumulation.

## Summary

- `F.scaled_dot_product_attention(q, k, v, attn_mask=...)` (or `is_causal=True`)
  returns numerically wrong output (no error, no NaN) once the query/key sequence
  length `S` exceeds `65536`.
- The default dispatch resolves to **`EFFICIENT_ATTENTION`** for this case
  (FLASH rejects the additive float mask; MATH would OOM at the real head count),
  and the default output matches forced-EFFICIENT bit-for-bit.
- `MATH` stays correct at all `S`.
- The error is a **step function**: `S = 65536` is fine (diff at bf16 rounding
  floor `1/128 = 0.0078`), `S = 65537` jumps to `~0.4–1.1` max-abs-diff vs MATH
  (exact value varies with the random draw).
- The threshold does **not** move with `head_dim` (64 / 128 / 256 all break at 65537).

## Environment

```
torch: 2.9.0a0+50eac811a6.nv25.09  (NVIDIA build)
cuda:  13.0
gpu:   NVIDIA H800 (sm_90), driver 570.133.20
dtype: bfloat16
```

## Minimal repro

```python
import torch, torch.nn.functional as F
from torch.nn.attention import sdpa_kernel, SDPBackend
import contextlib

dev, DT = torch.device("cuda"), torch.bfloat16
torch.manual_seed(0)

def sdpa(q, k, v, *, mask=None, causal=False, backend=None):
    ctx = sdpa_kernel([backend]) if backend is not None else contextlib.nullcontext()
    with torch.no_grad(), ctx:
        return F.scaled_dot_product_attention(q, k, v, attn_mask=mask, is_causal=causal)

import gc

# A) explicit additive (causal) bias, head_dim=128: which backend the DEFAULT dispatch
#    runs, and where it breaks. FLASH rejects the float mask; MATH is the reference.
print("== A: additive bias, head_dim=128 ==")
for S in [2048, 65535, 65536, 65537, 66005, 70000]:
    q, k, v = (torch.randn(1, 1, S, 128, device=dev, dtype=DT) for _ in range(3))
    tri = torch.triu(torch.ones(S, S, device=dev, dtype=torch.bool), 1)
    bias = torch.zeros(1, 1, S, S, device=dev, dtype=DT).masked_fill_(tri[None, None], torch.finfo(DT).min)
    ref = sdpa(q, k, v, mask=bias, backend=SDPBackend.MATH).float()
    row = {}
    for name, be in [("default", None), ("math", SDPBackend.MATH),
                     ("efficient", SDPBackend.EFFICIENT_ATTENTION), ("flash", SDPBackend.FLASH_ATTENTION)]:
        try:
            row[name] = f"{(sdpa(q,k,v,mask=bias,backend=be).float()-ref).abs().max().item():.5f}"
        except Exception as e:
            row[name] = f"({type(e).__name__})"
    print(S, row)
    del q, k, v, bias, tri, ref; gc.collect(); torch.cuda.empty_cache()

# B) head_dim independence. Uses is_causal (no materialized (S,S) bias) to keep the
#    sweep light; the additive-bias case in A breaks at the SAME S=65537, so the
#    boundary is independent of both head_dim and mask type.
print("== B: head_dim sweep, causal ==")
for HD in [64, 128, 256]:
    for S in [65536, 65537]:
        q, k, v = (torch.randn(1, 1, S, HD, device=dev, dtype=DT) for _ in range(3))
        ref = sdpa(q, k, v, causal=True, backend=SDPBackend.MATH).float()
        eff = sdpa(q, k, v, causal=True, backend=SDPBackend.EFFICIENT_ATTENTION).float()
        print(HD, S, f"{(eff-ref).abs().max().item():.5f}")
        del q, k, v, ref, eff; gc.collect(); torch.cuda.empty_cache()
```

## Observed output

**A. additive bias, head_dim=128 (default dispatch == efficient; MATH = 0; FLASH errors on the float mask):**

| S | >65536 | default | math | efficient | flash |
|---|---|---|---|---|---|
| 2048 | no | 0.00781 | 0.00000 | 0.00781 | RuntimeError |
| 65535 | no | 0.00781 | 0.00000 | 0.00781 | RuntimeError |
| 65536 | no | 0.00781 | 0.00000 | 0.00781 | RuntimeError |
| **65537** | YES | **0.87833** | 0.00000 | **0.87833** | RuntimeError |
| 66005 | YES | 0.98639 | 0.00000 | 0.98639 | RuntimeError |
| 70000 | YES | 1.08575 | 0.00000 | 1.08575 | RuntimeError |

**B. efficient-vs-math, sweeping head_dim (boundary fixed at 65537):**

| head_dim | S=65536 | S=65537 |
|---|---|---|
| 64 | 0.00391 | 0.46808 |
| 128 | 0.00781 | 0.51028 |
| 256 | 0.00781 | 0.60577 |

The exact max-abs-diff at S=65537 varies with the random draw (0.4–1.1 across runs);
what is invariant is the step from the bf16 floor (~0.004–0.008) at S<=65536 to ~0.5+ at
S=65537. A uses an explicit additive bias and B uses `is_causal`; both break at the same
`S=65537`, so the boundary is independent of mask type as well as head_dim.

## Why this looks like a 16-bit seq-index overflow, not numerical accumulation

1. **Discrete step, not gradual.** `S=65536` is at the bf16 rounding floor (0.0078);
   `S=65537` jumps to ~0.5–1.0 in one token. Numerical accumulation would grow
   smoothly with `S`, not step at a single index.
2. **Exactly 2^16.** The threshold is precisely 65536/65537, i.e. the first element
   whose 0-based index (65536) exceeds the `uint16` max (65535).
3. **Independent of head_dim.** 64/128/256 all break at the same `S=65537`. A
   combined offset like `seq * head_dim` would move the threshold with `head_dim`;
   a pure seq-dim index/stride would not. The latter matches.
4. **Silent and bounded.** No error, no NaN, output magnitude stays normal — i.e. a
   wrapped offset reads a wrong-but-valid location inside the same allocation, rather
   than an out-of-bounds illegal access (which would raise).

This points at a `int16`/`uint16`/`short` seq-dimension stride or offset inside the
efficient-attention (cutlass FMHA) kernel that wraps when the sequence index crosses
`2**16`. Note a separate failure mode exists for large **batch** (#142228) that DOES
raise (grid-dim launch error); the seq-length case here is silent, consistent with an
index wrap rather than a launch-config error.

## Relation to the original report

- The original report tested PyTorch **2.7.1**; this still reproduces on
  **2.9.0a0+nv25.09**, so the bug spans 2.7 through 2.9.
- The original report's data points (clean at S=50000, broken at S=100000) **straddle
  this 65536 boundary** (50000 < 65536 < 100000), i.e. they are consistent with a
  threshold at exactly 2^16 that the coarse 50K/100K spacing could not localize.
- The original "numerical accumulation" hypothesis does not fit the discrete step,
  the exact-2^16 threshold, or the head_dim independence shown above.

**Comment by drisspg** [comment]
This is fixed here: https://github.com/pytorch/pytorch/pull/187225
